kopf/clients/watching.py

```diff
@@ -83,51 +83,237 @@ def streaming_next(src):
+    """
+
+    # Tell the thread function that I/O errors are expected since now.
+    stopped_flag.set()
+
+    # If there is nothing to interrupt, do nothing.
+    if resp is None or not thread.is_alive():
+        return
+
+    # Simulate a disconnection. It is specially handled in the thread function.
+    if resp is not None and resp.raw:
+        if not resp.raw.closed:
+            os.close(resp.raw.fileno())
+
+    # Awaken the thread if it is in the blocking ``socket.read()`` call.
+    interrupted = __interrupt_thread(thread.ident)
+    if not interrupted:
+        logger.warning(
+            f"A watcher for {resource.name} cannot be interrupted. "
+            f"It can hang until a new event arrives. "
+            f"Use `configure_interruption()` for faster exiting.")
+
+    # Wait for the thread to finish: either due to I/O interruption,
+    # or due to socket timeout (if set), or by its own accord in other way.
+    await asyncio.get_running_loop().run_in_executor(None, thread.join)
+    if not interrupted:
+        logger.info(
+            f"A watcher for {resource.name} has finished on its own.")
+
+    # Re-raise if there were unexpected errors in the streaming thread.
+    thread_error.result(timeout=0)
+
+
+def _exec_streaming(src, loop, events_queue: asyncio.Queue, stopped_flag: asyncio.Event, thread_error: concurrent.futures.Future):
+    """
+    Stream by getting from a source and putting to a queue.
+
+    The exact type of the events is irrelevant, and is handled elsewhere.
+
+    This function is assumed to make blocking ``read()`` calls on sockets with
+    and without a timeout, which are later interrupted by `_stop_streaming`.
+    """
+    try:
+        # ``src`` comes from ``pykube.query.WatchQuery.object_stream()``.
+        for event in src:
+            future = asyncio.run_coroutine_threadsafe(events_queue.put(event), loop)
+            future.result()  # block, wait, re-raise.
+    except OSError as e:
+        # Only "bad file descriptor" is expected on graceful stream termination.
+        if e.errno == errno.EBADF and stopped_flag.is_set():
+            thread_error.set_result(None)
+        else:
+            thread_error.set_exception(e)
+    except BaseException as e:
+        thread_error.set_exception(e)
+        # raise  # prints to stderr, but otherwise ignores; useful for debugging
+    else:
+        thread_error.set_result(None)
+    finally:
+        # Wake up the queue consumer (otherwise it will hang on getting).
+        future = asyncio.run_coroutine_threadsafe(events_queue.put(None), loop)
+        future.result()  # block, wait, re-raise.
+
+
+#
+# Here comes the dark magic of low-level signals to interrupt threads & reads.
+#
+__interruption_lock = threading.Lock()
+__interruption_counter: int = 0
+__interruption_signal: Optional[int] = None
+__original_handler: Optional[Callable] = None
+
+
+def __can_interrupt():
+    return __interruption_signal is not None
+
+
```

Comment:
The number of underscores emphasises how dark this magic is, and how much hidden it should remain.